habitat/datasets/rearrange/sim_utilities.py

```diff
@@ -0,0 +1,838 @@
+    lowest_key_point: mn.Vector3 = None
+    lowest_key_point_height = None
+    highest_support_impact: mn.Vector3 = None
+    highest_support_impact_height = None
+    raycast_results = []
+    gravity_dir = sim.get_gravity().normalized()
+    object_local_to_global = obj.transformation
+    bb_corners = get_bb_corners(obj)
+    key_points = [mn.Vector3(0)] + bb_corners  # [COM, c0, c1 ...]
+    support_impacts: Dict[int, mn.Vector3] = {}  # indexed by keypoints
+    for ix, key_point in enumerate(key_points):
+        world_point = object_local_to_global.transform_point(key_point)
+        # NOTE: instead of explicit Y coordinate, we project onto any gravity vector
+        world_point_height = world_point.projected_onto_normalized(
+            -gravity_dir
+        ).length()
+        if (
+            lowest_key_point is None
+            or lowest_key_point_height > world_point_height
+        ):
+            lowest_key_point = world_point
+            lowest_key_point_height = world_point_height
+        # cast a ray in gravity direction
+        if ix == 0:
+            ray = habitat_sim.geo.Ray(world_point, gravity_dir)
+            raycast_results.append(sim.cast_ray(ray))
+            # classify any obstructions before hitting the support surface
+            for hit in raycast_results[-1].hits:
+                if hit.object_id == obj.object_id:
+                    continue
+                elif hit.object_id in support_obj_ids:
+                    hit_point = ray.origin + ray.direction * hit.ray_distance
+                    support_impacts[ix] = hit_point
+                    support_impact_height = mn.math.dot(
+                        hit_point, -gravity_dir
+                    )
+
+                    if (
+                        highest_support_impact is None
+                        or highest_support_impact_height
+                        < support_impact_height
+                    ):
+                        highest_support_impact = hit_point
+                        highest_support_impact_height = support_impact_height
+                # terminates at the first non-self ray hit
+                break
+    # compute the relative base height of the object from its lowest bb corner and COM
+    base_rel_height = (
+        lowest_key_point_height
+        - obj.translation.projected_onto_normalized(-gravity_dir).length()
+    )
+
+    surface_snap_point = (
+        None
+        if 0 not in support_impacts
+        else support_impacts[0] + gravity_dir * base_rel_height
+    )
+    # return list of obstructed and grounded rays, relative base height, distance to first surface impact, and ray results details
+    return {
+        "base_rel_height": base_rel_height,
+        "surface_snap_point": surface_snap_point,
+        "raycast_results": raycast_results,
+    }
+
+
+def snap_down(
+    sim: habitat_sim.Simulator,
+    obj: habitat_sim.physics.ManagedRigidObject,
+    support_obj_ids: Optional[List[int]] = None,
```

Comment:
We should explain what these "object ids" actually are... if I understand correctly, it's an id returned by our collision queries, and it can refer to either a rigid object, and articulated object base, or an articulated object link.